include/aws/sdkutils/private/endpoints_types_impl.h

```diff
@@ -28,84 +22,93 @@ enum aws_endpoints_expr_type {
     AWS_ENDPOINTS_EXPR_FUNCTION
 };
 
+enum aws_endpoints_fn_type {
+    AWS_ENDPOINTS_FN_FIRST = 0,
+    AWS_ENDPOINTS_FN_IS_SET = 0,
+    AWS_ENDPOINTS_FN_NOT,
+    AWS_ENDPOINTS_FN_GET_ATTR,
+    AWS_ENDPOINTS_FN_SUBSTRING,
+    AWS_ENDPOINTS_FN_STRING_EQUALS,
+    AWS_ENDPOINTS_FN_BOOLEAN_EQUALS,
+    AWS_ENDPOINTS_FN_URI_ENCODE,
+    AWS_ENDPOINTS_FN_PARSE_URL,
+    AWS_ENDPOINTS_FN_IS_VALID_HOST_LABEL,
+    AWS_ENDPOINTS_FN_AWS_PARTITION,
+    AWS_ENDPOINTS_FN_AWS_PARSE_ARN,
+    AWS_ENDPOINTS_FN_AWS_IS_VIRTUAL_HOSTABLE_S3_BUCKET,
+    AWS_ENDPOINTS_FN_LAST,
+};
+
 struct aws_endpoints_parameter {
     struct aws_allocator *allocator;
 
-    struct aws_byte_cursor name_cur;
-    struct aws_string *name;
+    struct aws_byte_cursor name;
 
-    enum aws_endpoints_value_type type;
-    struct aws_string *built_in;
+    enum aws_endpoints_parameter_value_type type;
+    struct aws_byte_cursor built_in;
 
+    bool has_default_value;
     union {
-        struct aws_string *string;
+        struct aws_byte_cursor string;
         bool boolean;
     } default_value;
 
     bool is_required;
-    struct aws_string *documentation;
+    struct aws_byte_cursor documentation;
     bool is_deprecated;
-    struct aws_string *deprecated_message;
-    struct aws_string *deprecated_since;
+    struct aws_byte_cursor deprecated_message;
+    struct aws_byte_cursor deprecated_since;
 };
 
 struct aws_endpoints_ruleset {
     struct aws_allocator *allocator;
     struct aws_ref_count ref_count;
 
+    struct aws_json_value *json_root;
+
     /* list of (aws_endpoints_rule) */
     struct aws_array_list rules;
 
-    struct aws_string *version;
-    struct aws_string *service_id;
+    struct aws_byte_cursor version;
+    struct aws_byte_cursor service_id;
     /* map of (aws_byte_cursor *) -> (aws_endpoints_parameter *) */
     struct aws_hash_table parameters;
 };
 
 struct aws_endpoints_function {
-    struct aws_string *fn; /* TODO: precompute hash or enum value to avoid lookup on every eval */
+    enum aws_endpoints_fn_type fn;
     /* List of (aws_endpoints_expr) */
     struct aws_array_list argv;
 };
 
 struct aws_endpoints_expr {
     enum aws_endpoints_expr_type type;
     union {
-        struct aws_string *string;
+        struct aws_byte_cursor string;
         double number;
         bool boolean;
         struct aws_array_list array; /* List of (aws_endpoints_expr) */
-        struct aws_string *reference;
+        struct aws_byte_cursor reference;
```

Comment:
the memory that these cursors are pointing to always outlives the `aws_endpoints_expr`?

and the memory is stable? it's not in an aws_byte_buf that could possibly resize or anything?

Reply:
yes. latest version keeps parsed cjson structure around. all the cursors point to strings inside of that json. lifetime for json is same as ruleset. ruleset keeps a couple byte_bufs around, but cursors never point to them. 
Will add comments on that.